Work around Composer multi-version loading hack using a "V2" namespace

tstarling opened the pull request:
Composer would try to load two versions of the same plugin, by reading the main plugin file and rewriting the class name. But this doesn't work if the plugin has multiple classes. It tries to run the new plugin class with old helper classes.

So:

* Move all classes to Wikimedia\Composer\Merge\V2. This includes classes that were previously in the parent namespace.
* Clean up unused "use" statements.

Note that the namespace version number will have to be incremented every time there is an internal interface change.

Closes #205

reedy commented on src/ExtraPackage.php:
```diff
@@ -346,7 +342,7 @@ protected function mergeConstraints(Link $origin, Link $merge, PluginState $stat
 
         if ($state->isComposer1()) {
-            $constraintClass = 'Wikimedia\\Composer\\Merge\\MultiConstraint';
+            $constraintClass = 'Wikimedia\\Composer\\Merge\\V2\\MultiConstraint';
```
Is it worth moving these to ::class (if they'll work) too?

Also, this looks kinda weird straight off; "if the state is composer (v) 1, use composer merge pluging V2 MultiConstraint, else use the one from Composer itself"

I know it's technically V2 of the plugin...

mcaskill replied:
I agree with you that the `V2` namespace feels weird. Sort of a necessary evil.

If only the Merge plugin had provided support for `setUpdateAllowList()` early on (before work on Composer 2 even began) then we _might_ have been able to avoid the renamespacing.

tstarling replied:
Done, used ::class. With ::class it's not necessary to reference V2 in this line of code.

mcaskill reviewed (commented):
The only alternative to this approach would be to implement runtime class replacement like Composer does for the main plugin file (see [`/src/Composer/Plugin/PluginManager.php`](https://github.com/composer/composer/blob/40095b2/src/Composer/Plugin/PluginManager.php#L204-L223)).

What if we introduced a `Factory` class that could do that, if it detects a key method is missing?
In our situation `ExtraPackage::getMergedRequirement()` or `PluginState::isComposer1()`.

tstarling commented:
I'm not really interested in exploring alternative approaches, I just want to release it and move on to the next thing. This way is simple, it works, and it's done already. If @mcaskill approves then I'll merge and release it as 2.0.0.

mcaskill commented on composer.json:
```diff
@@ -23,14 +23,14 @@
     },
     "extra": {
         "branch-alias": {
             "dev-master": "1.5.x-dev"
```
```suggestion
             "dev-master": "2.x-dev"
```

tstarling commented:
I tagged "2.0.0" before I realised that "v2.0.0" would be the correct convention for this package (for other packages I have used the plain version number without the "v"). So I deleted the tag, deleted the package version from packagist, and tagged v2.0.0. I didn't see v2.0.0 in packagist, so I released the same commit as v2.0.1, expecting that packagist got confused about the deleted version. Both v2.0.0 and v2.0.1 did end up in packagist eventually.

The point is, this is released as 2.0.1, please test and enjoy.
